<a href="https://colab.research.google.com/github/connerdandrews/cse380-notebooks/blob/master/07_2_Ponder_and_Prove_Elementary_Number_Theory.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Elementary Number Theory
#### Due: Saturday, 20 February 2021, 11:59 pm.

## Explore Fermat's Little Theorem Further


Fermat's Little Theorem (FLT) says that if $N$ is prime, then $N$ divides $X^N - X$.

Remember, the contrapositive of the conditional statement in this theorem is that if $N$ **doesn't** divide $X^N - X$ for some $X$, then $N$ **can't** be prime.

Unfortunately, this simple primality test doesn't always work, because it can be fooled by **pseudoprimes**.

For example, $341 = 11 \cdot 31$ is not prime. But $341$ **does** divide $2^{341} - 2$ as verified below:

In [ ]:
((2 ** 341) - 2) % 341

So $341$ is a so-called **base-2 pseudoprime**. What about **base-3**?

In [ ]:
((3 ** 341) - 3) % 341

Check that the result is not zero, therefore $341$ is **not** a **base-3 pseudoprime**.

Are there any other bases that will not fool the FLT test for $341$?

Are there any pseudoprimes that will fool the FLT for **any choice** of base coprime to them?

### The answer is yes.

Your task is the find the first 4-digit **bullet-proof pseudoprime** (**bppp**) and **prove** (yes, **PROVE**) that it will fool the FLT test for every base coprime to it.

Your proof must use all of the following:
1. the definition of coprime,
2. a consequence of coprimality,
3. the factorization of the **bppp**,
4. FLT, and the
5. CRT (Chinese Remainder Theorem).


In [ ]:
from math import gcd
from sympy import isprime

def passes_FLT_test_even_though_not_prime(b, n):
  # primes don't count as pseudoprimes
  return not isprime(n) and (b ** n) % n == b

def is_bppp(n):
  bases_coprime_to_n = [b for b in range(2, n) if gcd(b, n) == 1]
  return all(list(map(lambda b: passes_FLT_test_even_though_not_prime(b, n), 
                      bases_coprime_to_n)))

n = 1000
while not is_bppp(n):
  n += 1

n

1105

We now know by the calculation above that the first bullet proof pseudo-prime is 1105. What follows is a proof that 1105 truly is a bullet proof pseudo-prime. 

1105 is a 4-digit composite number. 

This is true because 1105 reduces into its prime factorization: 5 x 13 x 17

If two numbers are coprime than their greatest common denominator is 1. We can use the greatest common denominator of the prime factorization to ensure that 

gcd(b, 1105) = 1, gcd(b, 5) = 1, gcd(b, 13) = 1, gcd(b, 17) = 1



This means that if b is comprime to 1105 then it is coprime to 5, 13, and 17.

Fermat's Little Theorem tells us that b^(n-1) is congruent mod n to 1. 

Using this understanding we can then write the congruences that:
 
$b^{4} \equiv_{5} 1$

$b^{12} \equiv_{13} 1$

$b^{16} \equiv_{17} 1$
 
1104 is 1 less than 1105 and is a multiple of 4, 12, and 16. As a result the following also holds true. 

$(b^4)^{276} = b^{1104} \equiv_{5} 1$

$(b^{12})^{92} = b^{1104} \equiv_{13} 1$

$(b^{16})^{69} = b^{1104} \equiv_{17} 1$

Therefore, $b^{1104} \equiv_{1105} 1$

This works because the Chinese Remainder Theorem states that any two coprime numbers are congruent mod n to have the same remainder. 

1105 is a bullet proof pseudo-prime because it is coprime to 5, 13, and 17. As a result any n - 1 value would be equal to 4, 12, or 16. IF we factored those numbers into their prime factorizations, we would get 2 and 3.This means that 2 and 3 are congruent mod n to 1. This means that any number containing 2 or 3 in their prime factorizations would still be congruent mod n to 1. As a result, any number up to the square root of 1105 will still be congruent mod n to 1 and as a result will prove that 1105 is a pseudo-prime. 

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


# TODO My Report on What I Did and What I Learned

## Fun
I really enjoyed working on these problems. Even though I didn't fully understand every concept I felt that I enjoyed understanding the Chinese Remainder Theorem and the FMT. 

## New
I learned that the Chinese Remainder Theorem helps us to calculate the reminder when any number modulo n is calculated. This helped in my understanding of how we can determine of numbers are truly prime or if they are pseudo-primes. 

## Meaningful
I see that both the FMT and the Chinese Remainder Theorem are useful in helping us quickly determine if numbers are prime, co-prime, or pseudo-primes. This is useful when we need to find or use large numbers that we don't know if they are prime. This means that we can apply elementary number theory whenever we need large numbers in our calculations. 

## Other
Replace these words with your own describing other topics or sections of your report --- Connections, Collaborator Contributions, or anything else you feel impressed to add.

# TODO What is True?
Click on each warranted checkbox to toggle it to True (or back to False). 

NOTE: *This only works in Colab. If you run it in some other Jupyter notebook client/server environment you may have to change False to True (or vice versa) manually.*

This self-assessment is subject to revision by a grader.

In [5]:
#@markdown ## What is True about what I did?
#@markdown ### I had fun.
cb00 = True #@param {type:'boolean'}
#@markdown ### I learned something new.
cb01 = True #@param {type:'boolean'}
#@markdown ### I achieved something meaningful, or something I can build upon at a later time.
cb02 = True #@param {type:'boolean'}
#@markdown ## What is true about my report?
#@markdown ### I wrote a sufficient number of well-written sentences.
cb03 = True #@param {type:'boolean'}
#@markdown ### My report is free of mechanical infelicities.
cb04 = True #@param {type:'boolean'}
#@markdown ### I used Grammarly (or something better described in my report) to check for MIs.
cb05 = True #@param {type:'boolean'}
#@markdown ### I reported on any connections I found between these problems and something I already know. 
cb06 = True #@param {type:'boolean'}
#@markdown ### I reported who were and what contribution each of my collaborators made.
cb07 = True #@param {type:'boolean'}
#@markdown ## What is true about my proof?
#@markdown ### It succinctly uses the definition of coprime.
cb08 = True #@param {type:'boolean'}
#@markdown ### It correctly uses the definition of coprime.
cb09 = True #@param {type:'boolean'}
#@markdown ### It succinctly uses a consequence of comprimality
cb10 = False #@param {type:'boolean'}
#@markdown ### It correctly uses a consequence of comprimality
cb11 = False #@param {type:'boolean'}
#@markdown ### It succinctly uses the factorization of the **bppp**,
cb12 = True #@param {type:'boolean'}
#@markdown ### It correctly uses the factorization of the **bppp**,
cb13 = True #@param {type:'boolean'}
#@markdown ### It succinctly uses Fermat's Little Theorem. 
cb14 = True #@param {type:'boolean'}
#@markdown ### It correctly uses Fermat's Little Theorem. 
cb15 = True #@param {type:'boolean'}
#@markdown ### It succinctly uses the Chinese Remainder Theorem.
cb16 = True #@param {type:'boolean'}
#@markdown ### It correctly uses the Chinese Remainder Theorem.
cb17 = True #@param {type:'boolean'}



## DO NOT CHANGE ANYTHING IN THE NEXT CODE CELL!!
### Delete this cell and the following ones before submitting your work.

In [6]:
points_for_what_I_did = [5]*3
points_for_my_report = [7]*5
points_for_my_proof = [5]*10
points = points_for_what_I_did + points_for_my_report + points_for_my_proof
# cb is short for checkbox
total = sum(map(lambda n, p: p if eval(f'cb{n:02}') else 0,
                range(len(points)), points))             
total

90

# For graders

In [ ]:
#@markdown ---
number_of_MIs_found = 0 #@param {type: 'slider', min: 0, max: 5}
#@markdown ---
